src/hooks/useIsOnScreen.ts

```diff
@@ -0,0 +1,41 @@
+import { RefObject, useEffect, useRef, useState } from 'react';
+
+export const useIsOnScreen = (ref: RefObject<HTMLElement>) => {
+  const observerRef = useRef<IntersectionObserver | null>(null);
+  const [isOnScreen, setIsOnScreen] = useState(true);
+
+  useEffect(() => {
+    if (ref.current) {
+      const scrollParent = getScrollParent(ref.current);
+      observerRef.current = new IntersectionObserver(
+        ([entry]) => {
+          setIsOnScreen(entry.isIntersecting);
+        },
+        {
+          root: scrollParent === window ? undefined : (scrollParent as HTMLElement),
```

Comment:
won't this throw an exception in the SSR world?
`window` is oftentimes not defined there.

Reply:
```suggestion
          root: typeof window === 'undefined' || scrollParent === window ? undefined : (scrollParent as HTMLElement),
```
or
```suggestion
          root: scrollParent === globalThis.window ? undefined : (scrollParent as HTMLElement),
```
albeit I'm not sure whether the latter is supported well in older environments. Presumably not, so the first one is the safe bet

Reply:
Nope because it's in a `useEffect` - `useEffect` only runs in browser

Reply:
Yeah, I wasn't sure whether `useEffect` would be run or not, but I suppose it's just the sync one so the `useLayoutEffect`

Reply:
Yup!